File: solutions/java/rna-transcription/1/src/main/java/RnaTranscription.java
import java.util.Map;
import java.util.function.IntFunction;
import java.util.stream.Collectors;

class RnaTranscription {
	private static final Map<Character, Character> RNA_COMPLEMENT_BY_DNA = Map.of(
				'G', 'C',
				'C', 'G',
				'T', 'A',
				'A', 'U'
	);

    String transcribe(String dnaStrand) {
	    return dnaStrand.chars()
			           .mapToObj(this.getRnaCharacter())
			           .map(String::valueOf)
			           .collect(Collectors.joining());
    }

	private IntFunction<Character> getRnaCharacter() {
		return i -> getRna((char) i);
	}

	private char getRna(char dna) {
		Character rna = RNA_COMPLEMENT_BY_DNA.get(dna);
		if (rna == null) {
			throw new IllegalArgumentException("Invalid nucleotide: " + dna);
		}

		return rna;
	}

}

File: solutions/ocaml/rna-transcription/1/rna_transcription.ml
open Base

type dna = [ `A | `C | `G | `T ]
type rna = [ `A | `C | `G | `U ]

let transcribe : dna -> rna = function
    | `A -> `U
    | `C -> `G
    | `G -> `C
    | `T -> `A

let to_rna (strand : dna list) : rna list =
  List.map strand ~f:transcribe